### PR TITLE
feat: onboard new destination the trade desk

### PR DESCRIPTION
**Pull request by Gauravudia**
## What are the changes introduced in this PR?

Resolves [INT-897
](https://linear.app/rudderstack/issue/INT-897/firstpartydata-api-tradedesk)
- Onboard the new destination: `The Trade Desk`
- Onboard a destination on transformer proxy, addressing two key issues:
    - Handling error responses in the event of a `200` status code.
    - Resolving an issue caused by the server sorting payload keys, impacting the decryption of the HMAC at the Trade Desk's end. To overcome this challenge, a proxy request has been created, generating the HMAC of the sorted payload JSON keys.
- Used  new audience message type: `record`. This allows the server to send each record individually, with the transformer responsible for efficiently batching the events.
- Onboard destination on router to support new audience message type: record
- Added a utility function, `splitItemsBasedOnMaxSizeInBytes`, designed to split an items array containing various IDs (TDID, DAID, UID2) based on the maximum base size allowed by Trade Desk (approximately 2.5MB).
- Implemented a utility function, `getSignatureHeader`, to generate the HMAC of the payload using secret keys (advertiser secret key and provided data secret key).

## Please explain the objectives of your changes below

**Put down any required details on the broader aspect of your changes. If there are any dependent changes, mandatorily mention them here**

1. `Server changes to pass destination config in metadata is required for this Integration to work correctly. Without this change transformer will throw configuration error.`
2. `Adding data provider secret key in transformer env. `

### Type of change
`new feature`

### Any technical or performance related pointers to consider with the change?

N/A

### Any new dependencies introduced with this change?

N/A

### Any new generic utility introduced or modified. Please explain the changes.

existing utility: **prepareProxyRequest** in **network.js is** modified to return **config** from request. This config will be later used by calling function **proxyRequest** in **networkHandler.js** of **Trade Desk** 

### If the PR has changes in more than 10 files, please mention why the changes were not split into multiple PRs.

N/A

### If multiple linear tasks are associated with the PR changes, please elaborate on the reason:

N/A

<hr>

### Developer checklist

- [x] **No breaking changes are being introduced.**

- [x] Are all related docs linked with the PR?

- [x] Are all changes manually tested?

- [x] Does this change require any documentation changes?

- [x] Are relevant unit and component test-cases added?

### Reviewer checklist

- [x] Is the type of change in the PR title appropriate as per the changes?

- [x] Verified that there are no credentials or confidential data exposed with the changes.

### CODE REVIEW

File: src/adapters/network.js [hunk 1]
```diff
@@ -287,7 +287,7 @@ function getFormData(payload) {
  * @returns
  */
 const prepareProxyRequest = (request) => {
-  const { body, method, params, endpoint, headers } = request;
+  const { body, method, params, endpoint, headers, config } = request;
```
Comment (koladilip) [inline src/adapters/network.js:290]: should this be destinationConfig?

File: src/cdk/v2/destinations/the_trade_desk/utils.js
Comment (koladilip) [inline src/cdk/v2/destinations/the_trade_desk/utils.js:1]: why new files are not created in ts instead of js?

File: src/adapters/network.js (hunk 1 above)
Comment (Gauravudia) [reply src/adapters/network.js]: yes,

File: src/cdk/v2/destinations/the_trade_desk/utils.js [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+const getSignatureHeader = (request, secretKey) => {
+  if (!secretKey) {
+    throw new AbortedError('Secret key is missing. Aborting');
+  }
+  const sha1 = CryptoJS.HmacSHA1(JSON.stringify(request), secretKey);
+  const base = CryptoJS.enc.Base64.stringify(sha1);
+  return base;
+};
+
+const responseBuilder = (items, Config) => {
```
Comment (koladilip) [inline src/cdk/v2/destinations/the_trade_desk/utils.js:31]: we don't generally name args name starting with Capital letters

File: src/cdk/v2/destinations/the_trade_desk/utils.js [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+  });
+
+  if (currentBatch.length > 0) {
+    batches.push(currentBatch);
+  }
+
+  return batches;
+};
+
+const batchResponseBuilder = (items, Config) => {
```
Comment (koladilip) [inline src/cdk/v2/destinations/the_trade_desk/utils.js:73]: we don't generally name args name starting with Capital letters

File: src/cdk/v2/destinations/the_trade_desk/utils.js [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+  const payload = { DataProviderId: DATA_PROVIDER_ID, AdvertiserId: advertiserId, Items: items };
+
+  const response = defaultRequestConfig();
+  response.endpoint = getFirstPartyEndpoint(dataServer);
+  response.method = defaultPostRequestConfig.requestMethod;
+  response.body.JSON = removeUndefinedAndNullValues(payload);
+  return response;
+};
+
+const splitItemsBasedOnMaxSizeInBytes = (items, maxSize) => {
```
Comment (koladilip) [inline src/cdk/v2/destinations/the_trade_desk/utils.js:43]: https://github.com/rudderlabs/rudder-workflow-engine/blob/main/src/common/utils/batch.ts

this functionality is available here

File: src/cdk/v2/destinations/the_trade_desk/utils.js [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+
+  const response = getSuccessRespEvents(payloads, successMetadata, destination, true);
+  return [response, ...errorResponseList];
+};
+
+const processRouterDest = (inputs) => {
+  const respList = [];
+  const { destination } = inputs[0];
+  const groupedInputs = lodash.groupBy(inputs, (input) => input.message.type);
+  if (groupedInputs.record) {
```
Comment (koladilip) [inline src/cdk/v2/destinations/the_trade_desk/utils.js:122]: are we expecting only record type events?

File: src/v0/destinations/the_trade_desk/networkHandler.js [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+  processAxiosResponse,
+  getDynamicErrorType,
+} = require('../../../adapters/utils/networkUtils');
+const { getSignatureHeader } = require('../../../cdk/v2/destinations/the_trade_desk/utils');
+const { isHttpStatusSuccess } = require('../../util/index');
+const tags = require('../../util/tags');
+const { JSON_MIME_TYPE } = require('../../util/constant');
+
+const proxyRequest = async (request) => {
+  console.log('request', JSON.stringify(request));
```
Comment (koladilip) [inline src/v0/destinations/the_trade_desk/networkHandler.js:13]: remove this

File: src/cdk/v2/destinations/the_trade_desk/utils.js (hunk 5 above)
Comment (Gauravudia) [reply src/cdk/v2/destinations/the_trade_desk/utils.js:122]: yes

File: src/cdk/v2/destinations/the_trade_desk/utils.js (hunk 2 above)
Comment (Gauravudia) [reply src/cdk/v2/destinations/the_trade_desk/utils.js]: updated, used **config**

File: src/cdk/v2/destinations/the_trade_desk/utils.js (hunk 3 above)
Comment (Gauravudia) [reply src/cdk/v2/destinations/the_trade_desk/utils.js]: updated, used **config**

File: src/v0/destinations/the_trade_desk/networkHandler.js (hunk 6 above)
Comment (Gauravudia) [reply src/v0/destinations/the_trade_desk/networkHandler.js]: removed

**Review by yashasvibajpai — APPROVED** [review]

File: src/v0/destinations/the_trade_desk/networkHandler.js [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+  getDynamicErrorType,
+} = require('../../../adapters/utils/networkUtils');
+const { getSignatureHeader } = require('../../../cdk/v2/destinations/the_trade_desk/util');
+const { isHttpStatusSuccess } = require('../../util/index');
+const tags = require('../../util/tags');
+const { JSON_MIME_TYPE } = require('../../util/constant');
+
+const getSecretKeyFromMetadata = (metadata, secretKey) => {
+  const { secret } = metadata[0];
+  if (!secret?.[secretKey]) {
```
[the remark this answers is not in the document]
Comment (Gauravudia) [reply src/v0/destinations/the_trade_desk/networkHandler.js]: This utility is remove and we are not using it. Can you refresh the PR and do the fresh review?

File: src/cdk/v2/destinations/the_trade_desk/utils.js
Comment (Gauravudia) [reply src/cdk/v2/destinations/the_trade_desk/utils.js:1]: I will address this in a separate PR.

File: src/v0/destinations/the_trade_desk/networkHandler.js [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+  };
+
+  const requestOptions = {
+    url: endpoint,
+    data,
+    params,
+    headers: ProxyHeaders,
+    method,
+  };
+  const response = await httpSend(requestOptions, { feature: 'proxy', destType: 'the_trade_desk' });
```
[the remark this answers is not in the document]
Comment (Gauravudia) [reply src/v0/destinations/the_trade_desk/networkHandler.js:40]: endpoint is coming from proxy request from server.

[the remark this answers is not in the document]
Comment (Gauravudia) [reply src/v0/destinations/the_trade_desk/networkHandler.js:40]: what it use of endpoint path here?

**Review by koladilip — DISMISSED** [review]

**Comment by koladilip** [comment]
so we can merge this directly or we need to deploy this only to exinity? cc: @krishna2020

File: src/v0/destinations/the_trade_desk/networkHandler.js [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+
+const proxyRequest = async (request) => {
+  const { endpoint, data, method, params, headers, config } = prepareProxyRequest(request);
+
+  if (!config?.advertiserSecretKey) {
+    throw new AbortedError('Advertiser secret key is missing in destination config. Aborting');
+  }
+
+  if (!process.env.THE_TRADE_DESK_DATA_PROVIDER_SECRET_KEY) {
+    throw new AbortedError('Data provider secret key is missing. Aborting');
```
Comment (krishna2020) [inline src/v0/destinations/the_trade_desk/networkHandler.js:20]: These should be Instrumentation Errors as customer misconfigured the destination

**Comment by krishna2020** [comment]
> so we can merge this directly or we need to deploy this only to exinity? cc: @krishna2020

Requirement is only for exinity customer. This PR has changes only w.r.t TradeDesk destination - so good to merge too. However exinity should have dedicated transformer deployment for enabling v1 proxy - so deploying this branch also works.

File: src/v0/destinations/the_trade_desk/networkHandler.js (hunk 9 above)
Comment (Gauravudia) [reply src/v0/destinations/the_trade_desk/networkHandler.js:20]: it is coming from environment variable.

**Review by yashasvibajpai — DISMISSED** [review]

File: src/v0/destinations/the_trade_desk/networkHandler.js (hunk 9 above)
Comment (Gauravudia) [reply src/v0/destinations/the_trade_desk/networkHandler.js]: plaformError is the correct class we can use here.

**Review by koladilip — APPROVED** [review]

File: test/integrations/destinations/the_trade_desk/delivery/data.ts [hunk 10]
```diff
@@ -0,0 +1,248 @@
...
+  sampleDestination,
+} from '../common';
+
+beforeAll(() => {
+  process.env.THE_TRADE_DESK_DATA_PROVIDER_SECRET_KEY = 'mockedDataProviderSecretKey';
+});
+
+afterAll(() => {
+  delete process.env.THE_TRADE_DESK_DATA_PROVIDER_SECRET_KEY;
+});
```
Comment (sanpj2292) [inline test/integrations/destinations/the_trade_desk/delivery/data.ts:16]: Not sure if `afterAll` & `beforeAll` make sense here. This is more of a `P1` issue or Non-blocking comment

**Review by yashasvibajpai — APPROVED** [review]

**Pull request merged**